src/go/nerdctl-stub/command_handlers.go

```diff
@@ -22,15 +25,133 @@ func imageBuildHandler(c *commandDefinition, args []string) (*parsedArgs, error)
+	// the container or outside.
+
+	for _, arg := range args {
+		if arg == "-" || !strings.HasPrefix(arg, "-") {
+			// If the arg is "-" (stdin/stdout) or doesn't start with -, it's a path.
+			paths = append(paths, arg)
+		} else {
+			resultArgs = append(resultArgs, arg)
+		}
+	}
+
+	if len(paths) != 2 {
+		// We should have exactly one source and one destination... just fail
+		err := fmt.Errorf("accepts 2 args, received %d", len(paths))
+		if cleanupErr := runCleanups(cleanups); cleanupErr != nil {
+			err = multierror.Append(err, cleanupErr)
+		}
+		return nil, err
+	}
+
+	hostPathDeterminerFuncs := []func(i int, p string) hostPathResult{
+		func(i int, p string) hostPathResult {
+			if p == "-" {
+				// If one argument is "-", the other must be a container path, so
+				// neither needs to be modified.
+				return hostPathNeither
+			}
+			return hostPathUnknown
+		},
+		func(i int, p string) hostPathResult {
+			colon := strings.Index(p, ":")
+			if colon < 1 {
+				// If there's no colon in the path specification at all, or if the
+				// string starts with a colon (which is invalid), then this must not be
+				// a container path (and therefore the other one is).
+				return hostPathCurrent
+			}
+			return hostPathUnknown
+		},
+		func(i int, p string) hostPathResult {
+			colon := strings.Index(p, ":")
+			if colon > 1 {
+				// There's multiple characters before the first colon; this is a container
+				// path specification (foo:/path/in/container), so the other must be a
+				// host path specification.
+				return hostPathOther
+			}
+			return hostPathUnknown
+		},
+		func(i int, p string) hostPathResult {
+			colon := strings.Index(p, ":")
+			if colon != 1 {
+				// Shouldn't get here -- one of the two previous functions should have
+				// found something already.
+				panic(fmt.Sprintf("Expected path %q to start with a character followed by a colon!", p))
+			}
+			// Fall back: the first element should be treated as the container path.
+			if i == 0 {
+				return hostPathOther
+			}
+			return hostPathCurrent
+		},
+	}
+
+functionLoop:
+	for _, f := range hostPathDeterminerFuncs {
+		for i, p := range paths {
+			result := f(i, p)
+			hostPathIndex := i
+			switch result {
+			case hostPathNeither:
+				resultArgs = append(resultArgs, paths...)
+				break functionLoop
+			case hostPathUnknown:
+				continue
+			case hostPathOther:
+				hostPathIndex = 1 - i
+			}
+
+			// If we reach here, we found the host path to munge.
+			if hostPathIndex == 1 {
+				resultArgs = append(resultArgs, paths[0])
+			}
+			newPath, newCleanups, err := argHandlers.filePathArgHandler(paths[hostPathIndex])
+			if err != nil {
+				cleanups = append(cleanups, newCleanups...)
+				if cleanupErr := runCleanups(cleanups); cleanupErr != nil {
+					err = multierror.Append(err, cleanupErr)
+				}
+				return nil, err
+			}
+			resultArgs = append(resultArgs, newPath)
+			if hostPathIndex != 1 {
+				resultArgs = append(resultArgs, paths[1])
+			}
```

Comment:
Wouldn't this be the same as:

```suggestion
			paths[hostPathIndex], newCleanups, err := argHandlers.filePathArgHandler(paths[hostPathIndex])
			if err != nil {
				cleanups = append(cleanups, newCleanups...)
				if cleanupErr := runCleanups(cleanups); cleanupErr != nil {
					err = multierror.Append(err, cleanupErr)
				}
				return nil, err
			}
			resultArgs = append(resultArgs, paths...)
```

Reply:
Yes, that looks much clearer.